src/dns_server.c

```diff
@@ -663,32 +663,42 @@ static int _dns_setup_ipset(struct dns_request *request)
 
 	/* check ipset rule */
 	rule_flags = request->domain_rule.rules[DOMAIN_RULE_FLAGS];
-	if (rule_flags) {
-		if (rule_flags->flags & DOMAIN_FLAG_IPSET_IGNORE) {
-			return 0;
-		}
-	}
-
-	ipset_rule = request->domain_rule.rules[DOMAIN_RULE_IPSET];
-	if (ipset_rule == NULL) {
-		return 0;
-	}
 
 	/* add IPV4 to ipset */
-	if (request->has_ipv4 && request->qtype == DNS_T_A) {
+	if (!(rule_flags && (rule_flags->flags & DOMAIN_FLAG_IPSET_IPV4_IGN)) && request->has_ipv4 &&
+		request->qtype == DNS_T_A) {
+		ipset_rule = request->domain_rule.rules[DOMAIN_RULE_IPSET_IPV4];
+		if (ipset_rule == NULL) {
+			/* fallback to ipv6 */
+			ipset_rule = request->domain_rule.rules[DOMAIN_RULE_IPSET_IPV6];
+			if (ipset_rule == NULL) {
+				return 0;
+			}
+		}
+		if (ipset_rule == NULL) {
+			return 0;
+		}
```

Comment:
The function is a little complex, and not easy to read, please consider the following suggestion code.
code not tested.

```suggestion
static int _dns_setup_ipset(struct dns_request *request)
{
	struct dns_ipset_rule *ipset_rule = NULL;
	struct dns_rule_flags *rule_flags = NULL;
	int ret = 0;

	if (_dns_server_has_bind_flag(request, BIND_FLAG_NO_RULE_IPSET) == 0) {
		return 0;
	}

	/* check ipset rule */
	rule_flags = request->domain_rule.rules[DOMAIN_RULE_FLAGS];

	switch (request->qtype) {
	case DNS_T_A: {
		if (rule_flags) {
			if (rule_flags->flags & DOMAIN_FLAG_IPSET_IPV4_IGN) {
				return 0;
			}
		}

		ipset_rule = request->domain_rule.rules[DOMAIN_RULE_IPSET_IPV4];
		if (ipset_rule == NULL) {
			return 0;
		}

		/* add IPV4 to ipset */
		if (request->has_ipv4 && request->qtype == DNS_T_A) {
			ret |= ipset_add(ipset_rule->ipsetname, request->ipv4_addr, DNS_RR_A_LEN, request->ttl_v4 * 2);
		}
	} break;
	case DNS_T_AAAA: {
		if (rule_flags) {
			if (rule_flags->flags & DOMAIN_FLAG_IPSET_IPV6_IGN) {
				return 0;
			}
		}

		ipset_rule = request->domain_rule.rules[DOMAIN_RULE_IPSET_IPV6];
		if (ipset_rule == NULL) {
			return 0;
		}

		/* add IPV6 to ipset */
		if (request->has_ipv6 && request->qtype == DNS_T_AAAA) {
			if (request->has_ipv4) {
				ret |= ipset_add(ipset_rule->ipsetname, request->ipv4_addr, DNS_RR_A_LEN, request->ttl_v4 * 2);
			}
			ret |= ipset_add(ipset_rule->ipsetname, request->ipv6_addr, DNS_RR_AAAA_LEN, request->ttl_v6 * 2);
		}
	} break;
	default:
		return ret;
	}

	tlog(TLOG_DEBUG, "IPSET-MATCH: domain:%s, ipset:%s, result: %d", request->domain, ipset_rule->ipsetname, ret);

	return ret;
}

```

Reply:
I don't know whether there will be IPv4 result when `qtype == DNS_T_AAAA`. If so, rules like `ipset /example.com/#4:-,#6:dns4` will fail when `dns4` is an IPv4 ipset. If not so, I don't see the point to process IPv4 address in IPv6 branch.
Besides, there's no fallback when one stack is set while the other is not.

Out of code reusability, maybe a new function is better, but it might have too many parameters.

Reply:
Any reply?

Reply:
Three suggestion:
1. Reduce the complexity of the function.
2. When requesting AAAA, if there is an IPV4 address, and not set to ignore IPV4 rules, ipset should be set.
3. When requesting A records, no fallback IPV6 is required.

Reply:
Any progress?

Reply:
@hcwang I have been busy in assignments recently. Will update ASAP.

Reply:
Thanks and waiting for the good news.